src/components/SearchInput/SearchInput.tsx

```diff
@@ -0,0 +1,108 @@
+import {
+  ChangeEvent,
+  ComponentProps,
+  FocusEvent,
+  MouseEvent,
+  ReactNode,
+  useEffect,
+  useState,
+} from "react"
+import styles from "./SearchInput.module.css"
+
+interface SearchInputProps extends ComponentProps<"input"> {
+  searchIcon?: ReactNode
+  clearIcon?: ReactNode
+  onDebounce?: (searchTerm: string) => void
+  debounceDelay?: number
+  onSearchClick?: (searchTerm: string) => void
+  value?: string
+}
```

Comment:
`onDebounce` feels like Leaky Abstraction, so does `onSearchClick`. 

Debounce is a technique for dealing with event dispatching, preventing multiple events to be dispatched in a short amount of time.

Whether by user clicking or a debounce timing out, the important thing about this component is that it triggers an search action with some `(searchTerm: string)` (notice how it shows up twice in the interface), so it would be expected in my opinion to see a `onSearch` callback prop that _could_ be triggered by click or debounce, and the beauty of it is that the consumer of this component doesn't have to know/worry about clicks and debounces.

Reply:
@luiz-chagas so I moved forward with this advice, however, I think it loses the convenience of avoiding the callback being fired until the button is clicked. Trying to think of the best strategy here. Do we maybe just not debounce when a the `showSearchButton` boolean is `true`?

Reply:
My other thought was returning the event type in the callback so the developer could programmatically decided how to handle each type.

Reply:
I'm leaning towards the former. The latter seems unnecessarily complex and at that point you might as well have two different callbacks.